hpy/devel/src/runtime/ctx_type.c

```diff
@@ -0,0 +1,212 @@
+#include <stddef.h>
+#include <Python.h>
+#include "hpy.h"
+#include "common/runtime/ctx_type.h"
+
+#ifdef HPY_UNIVERSAL_ABI
+   // for _h2py and _py2h
+#  include "handles.h"
+#endif
+
+/* by default, the C structs which bake an HPy custom type do NOT include
+ * PyObject_HEAD.  So, HPy_New must allocate a memory region which is big
+ * enough to contain PyObject_HEAD + any eventual extra padding + the actual
+ * user struct. We use the union_alignment to ensure that the payload is
+ * correctly aligned for every possible struct.
+ */
+
+typedef struct {
+    PyObject_HEAD
+    union {
+        unsigned char payload[1];
+        // these fields are never accessed: they are present just to ensure
+        // the correct alignment of payload
+        unsigned short _m_short;
+        unsigned int _m_int;
+        unsigned long _m_long;
+        unsigned long long _m_longlong;
+        float _m_float;
+        double _m_double;
+        long double _m_longdouble;
+        void *_m_pointer;
+    };
+} GenericHPyObject;
+
+#define PyObject_HEAD_SIZE (offsetof(GenericHPyObject, payload))
+
+
+static void methoddef_copy_ml_meth_and_flags(HPyMethodDef *src, PyMethodDef *dst)
+{
+#ifdef HPY_UNIVERSAL_ABI
+    // Universal ABI mode: ml_meth could be either a legacy function or a
+    // new-style HPy method
+    if (src->ml_flags & _HPy_METH) {
+        // HPy function: cal ml_meth to get pointers to the impl_func and
+        // the cpy trampoline
+        void *impl_func;
+        PyCFunction trampoline_func;
+        src->ml_meth(&impl_func, &trampoline_func);
+        dst->ml_meth = (PyCFunction)trampoline_func;
+        dst->ml_flags = src->ml_flags & ~_HPy_METH;
+    }
+    else {
+        // legacy function: ml_meth already contains a function pointer
+        // with the correct CPython signature
+        dst->ml_meth = (PyCFunction)src->ml_meth;
+        dst->ml_flags = src->ml_flags;
+    }
+#else
+    // CPython ABI mode: ml_meth is already a PyCFunction function pointer
+    dst->ml_meth = src->ml_meth;
+    dst->ml_flags = src->ml_flags;
+#endif
+}
+
+static void slot_copy_pfunc(HPyType_Slot *src, PyType_Slot *dst)
+{
+#ifdef HPY_UNIVERSAL_ABI
+    // Universal ABI mode: for now we assume that it uses the HPy calling
+    // convention, but we need a way to allow slots with the CPython
+    // signature, similar to what we do in methoddef_copy_ml_meth_and_flags
+    HPyMeth f = (HPyMeth)src->pfunc;
+    void *impl_func;
+    _HPy_CPyCFunction trampoline_func;
+    f(&impl_func, &trampoline_func);
+    dst->pfunc = trampoline_func;
+#else
+    // CPython ABI mode
+    dst->pfunc = src->pfunc;
+#endif
+}
+
+
+_HPy_HIDDEN void*
+ctx_Cast(HPyContext ctx, HPy h)
+{
+    // XXX: how do we implement ctx_Cast when has_pyobject_head==1?
```

Comment:
For now we should check the type object to see if has_pyobject_head==1 or not.  It doesn't seem like a big overhead to me.  We can store that info in a quickly-accessible place, like there are probably tp_flags bits we can abuse.

Reply:
Yes, I was also thinking about (ab)using `tp_flags`.
What I don't like is that in theory `HPy_CAST` was supposed to be a purely static cast for the sake of the typesystem, now it becomes something which has a runtime behavior, with a bit of overhead (although I admit it's probably very small).

But I suppose that for now it's fine.